File: native-apps/desktop/macos/Core/Network/WorkAreaSession.swift
import Foundation
import WebKit

/// Native нэвтрэлтийг ажлын муж руу дамжуулагч.
///
/// `APIClient` нь cookie-г ЗОРИУДААР хүлээж авдаггүй (`httpCookieAcceptPolicy
/// = .never`) бөгөөд тэр нь зөв хэвээр: native дуудлагууд session биш,
/// `sessionId` + `pollToken` хосоор ажилладаг тул ambient authority шаардахгүй.
///
/// Гэвч платформын нэвтрэлт (`/api/v1/auth/eid/poll`) амжилттай болоход сервер
/// `session_token`-ыг `Set-Cookie`-гээр буцаадаг ба тэр нь **ажлын мужийн
/// цорын ганц нэвтрэлт**: төхөөрөмжийн шугам дээр web-ийн `/login` хаалттай
/// (`frontend/proxy.ts` — тэнд нэвтрэлт бол native UI) тул webview өөрөө
/// нэвтэрч ЧАДАХГҮЙ. Энэ дамжуулалтгүйгээр ажлын муж нь native талд нэвтэрсэн
/// хүнийг ч «нэвтрээгүй» гэж уншина — яг тэр байдал 2026-09-03 хүртэл
/// үргэлжилсэн.
///
/// Cookie-г задлаад дахин угсрахгүй, серверийн толгойг `HTTPCookie`-оор
/// шууд уншина: `HttpOnly`, `Secure`, `SameSite`, дуусах хугацаа бүгд хэвээр
/// үлдэнэ. Гараар барьсан cookie нь тэдгээрийг чимээгүй унагаадаг.
enum WorkAreaSession {

    /// `backend/internal/kernel/security/csrf.go` → `TenantSessionCookie`.
    static let cookieName = "session_token"

    /// Хариунд session cookie байвал webview-ийн сан руу суулгана.
    ///
    /// Бусад бүх cookie-г алгасна: энэ нь cookie jar биш, нэг тодорхой
    /// нэвтрэлтийн гүүр.
    static func capture(from response: HTTPURLResponse, for url: URL?) {
        guard let url,
              let fields = response.allHeaderFields as? [String: String] else { return }
        let session = HTTPCookie.cookies(withResponseHeaderFields: fields, for: url)
            .filter { $0.name == cookieName && !$0.value.isEmpty }
        guard !session.isEmpty else { return }
        Task { @MainActor in
            for cookie in session {
                await WKWebsiteDataStore.default().httpCookieStore.setCookie(cookie)
            }
        }
    }

    /// Гарахад ажлын мужийг мөн гаргана. Native тал гарчихаад webview нь
    /// нэвтэрсэн хэвээр үлдэх нь дараагийн хүнд өмнөх хүний мужийг үзүүлнэ.
    static func clear() {
        Task { @MainActor in
            let store = WKWebsiteDataStore.default().httpCookieStore
            for cookie in await store.allCookies() where cookie.name == cookieName {
                await store.deleteCookie(cookie)
            }
        }
    }
}

File: native-apps/mobile/android/app/src/main/kotlin/mn/petronet/app/net/WorkAreaSession.kt
package mn.petronet.app.net

import android.webkit.CookieManager
import mn.petronet.app.AppConfig

/**
 * Native нэвтрэлтийг ажлын муж руу дамжуулагч — iOS-ийн `WorkAreaSession.swift`-ийн дүйцэл.
 *
 * OkHttp нь анхдагчаараа cookie хадгалдаггүй (`CookieJar.NO_COOKIES`) бөгөөд
 * тэр нь зөв хэвээр: native дуудлагууд session биш, `sessionId` + `pollToken`
 * хосоор ажилладаг. Гэвч платформын нэвтрэлт (`/api/v1/auth/eid/poll`)
 * амжилттай болоход сервер `session_token`-ыг `Set-Cookie`-гээр буцаадаг ба
 * тэр нь **ажлын мужийн цорын ганц нэвтрэлт**: төхөөрөмжийн шугам дээр
 * web-ийн `/login` хаалттай (`frontend/proxy.ts`) тул WebView өөрөө нэвтэрч
 * ЧАДАХГҮЙ.
 *
 * Толгойг задлаад дахин угсрахгүй, `CookieManager` руу түүхийг нь ХЭВЭЭР
 * дамжуулна: `HttpOnly`, `Secure`, `SameSite`, дуусах хугацаа бүгд үлдэнэ.
 */
object WorkAreaSession {

    /** `backend/internal/kernel/security/csrf.go` → `TenantSessionCookie`. */
    private const val COOKIE_NAME = "session_token"

    /** `Set-Cookie` толгой бүрд нэг удаа. Бусад cookie-г алгасна. */
    fun capture(setCookieHeader: String) {
        if (!setCookieHeader.startsWith("$COOKIE_NAME=")) return
        CookieManager.getInstance().apply {
            setAcceptCookie(true)
            setCookie(AppConfig.baseUrl, setCookieHeader)
            flush()
        }
    }

    /**
     * Гарахад ажлын мужийг мөн гаргана — эс бөгөөс дараагийн хүн өмнөх хүний
     * мужийг табаа дарахад л харна.
     *
     * ponytail: тухайн нэг cookie-г нэрээр нь устгах API байхгүй тул хугацааг
     * нь өнгөрүүлж дарна — бүх cookie-г цэвэрлэвэл хүний өөрийн хэл, харагдац
     * зэрэг ажлын мужийн бусад тохиргоо хамт алга болно.
     */
    fun clear() {
        CookieManager.getInstance().apply {
            setCookie(AppConfig.baseUrl, "$COOKIE_NAME=; Path=/; Max-Age=0")
            flush()
        }
    }
}
